zip-0308.rst

```diff
@@ -0,0 +1,326 @@
+
+Some wallets by default add a "developer fee" to each transaction, directed
+to the developer(s) of the wallet. This is typically implemented by adding
+the developer address as an explicit output, so if migration transactions are
+generated internally by ``zcashd``, they will not include the developer fee.
+We strongly recommend *not* patching the ``zcashd`` code to add the developer
+fee output to migration transactions, because doing so partitions the anonymity
+set between users of that wallet and other users.
+
+There MUST NOT be any transparent inputs or outputs, or Sapling Spends, in a
+migration transaction.
+
+The ``lock_time`` field MUST be set to 0 (unused).
+
+
+Open questions
+--------------
+
+The above strategy has several "magic number" parameters:
+
+* the interval between batches (500 blocks)
+* the maximum number of transactions in a batch (5)
+* the distribution of exponents (uniform integer in 6..9)
+* the distribution of mantissae (uniform integer in 1..99).
+
+These have been chosen by guesswork. Should we change any of them?
+
+In particular, if the amount to migrate is large, then this strategy can
+result in large amounts (up to 990 ZEC, worth USD ~49500 at time of writing)
+transferred in each transaction. This leaks the fact that the transaction was
+sent by a user who has at least that amount.
+
+The strategy does not migrate any remaining fractional amount less than
+0.01 ZEC (worth USD ~0.50 at time of writing). Is this reasonable?
+
+In deciding the amount to send in each transaction, the strategy does not
+take  account of the values of individual Sprout notes, only the total amount
+remaining to migrate. Can a strategy that is sensitive to individual note
+values improve privacy?
+
+An adversary may attempt to interfere with the view of the block chain seen
+by a subset of nodes that are performing migrations, in order to cause those
+nodes to send migration batches at a different time, so that they may be
+distinguished. Is there anything further we can do to mitigate this
+vulnerability?
+
+
+RPC calls
+---------
+
+Nodes MUST maintain a boolean state variable recording whether migration is
+enabled. This can be set or reset using the following RPC call::
+
+  setenablemigration true|false
+
+It is intentional that there is no facility to change options associated with
+the migration, since such options can distinguish users.
+
+Open questions:
+
+* Should the state variable be persisted, and how?
+* Is there also a config option?
+
+
+Nodes MUST also support the following RPC call to return the current status of
+the migration::
+
+  getmigrationstatus
+
+Returns::
+
+  {
+    "enabled": true|false,
+    "unmigrated_amount": nnn.n,
+    "unconfirmed_migrated_amount": nnn.n,
+    "confirmed_migrated_amount": nnn.n,
+    "confirmed_migration_transactions": nnn,
+    "time_started": ttt, // Unix timestamp
+    "estimated_time_remaining_seconds": ttt
+  }
```

Comment:
Maybe the `getmigrationstatus` RPC can also return a list of TxIDs that are a part of this migration. It'll be nice to show this in the UI.

Reply:
This was added: `"migration_txids": [txids]`